luals-addon/factorio/factorio-plugin/util.lua

```diff
@@ -170,6 +170,192 @@ local function add_chain_diff(chain_diff, diffs)
   end
 end
 
+---@alias SourceRange {from: integer, to: integer, length: integer, content: string}
+---@alias LexerState
+--- | '"code"'
+--- | '"short_string"'
+--- | '"long_string"'
+--- | '"short_comment"'
+--- | '"long_comment"'
+
+---Lexically analyze Lua source files for positions of strings and comments.
+---Notably, this needs to be able to handle 'long brackets', which are context-sensitive.
+---We should really only be doing this once per source file.
+---@param source string
+---@return SourceRange[]
+local function lex_lua_nonexecutables(source)
+  ---@type SourceRange[]
+  local ranges = {}
+  ---@type LexerState
+  local state = "code"
+  local cursor = 1 -- 1 is the first character in the source file.
+
+  local delimit = ""
+  local patterned_delimit = ""
+  local start = 0
+  local _end = 0
+
+  local char_escaped = false
+
+  ---check if the next character(s) are equal to the given string
+  ---@param query string
+  ---@return boolean
+  local function peek(query)
+    if cursor + #query > #source then return false end
+    return source:sub(cursor, cursor + #query - 1) == query
+  end
+
+  local function take(query)
+    if not peek(query) then return false end
+    cursor = cursor + #query
+    return true
+  end
+
+  ---Parse a opening long bracket, like `[[` or `[=[`
+  ---Assumes the first bracket has already been consumed.
+  ---@return boolean, integer | nil
+  local function parse_longbracket_open()
+    -- Consume all the '='s
+    local count = 0
+    while take("=") do
+      count = count + 1
+    end
+    if take("[") then
+      return true, count
+    else
+      return false, nil
+    end
+  end
+
+  local function append_range()
+    ranges[#ranges + 1] = {
+      from = start,
+      to = _end,
+      length = _end - start + 1,
+      content = source:sub(start, _end),
+    }
```

Comment:
This could be turned into an array: `{start, _end}`
(or maybe just have `ranges` be `{start1, end1, start2, end2, ..., startN, endN}`)

`length` and `content` are not actually used anywhere outside of testing and might cause slowdown with interning.

Reply:
Yep, it would be the latter. Reason being that the latter does not do a heap allocation (table construction) for every range, it's just 1 big table. And that table can remain large, so by reusing the same table throughout the lifetime of the entire plugin it cuts down on allocations significantly. And allocations are as per usual the expensive part in programming.